README.md
DDI Search
==========

Django website for search and display of Data Documentation Initiative (DDI) XML


CHANGELOG.rst
.. _CHANGELOG:

CHANGELOG
=========

The following is a summary of changes and improvements to the Finding
Aids application.  New features in each version should be listed, with the most
recent version first.  Upgrade or deployment notes should be found in
:ref:`DEPLOYNOTES`.

1.1.1
-----

* Enable Google Analytics

1.1
---

* Upgrade to eXist-db 2.2
* Upgrade to Django 1.9
* bugfix: error handling for unrecognized geonames id when loading documents
* Configured for testing on Travis-CI
* Configured to use django-debug-toolbar with eulexistdb for development

1.0.1
-----

* bugfix: DOI detection now also works on DOI URLs in the bibliographic citation.

1.0
---

Search & Browse Functionality
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

* As a researcher, I want to search by keyword or phrase across all data collections
  as a simple way to find records that include all of my search terms.
* As a researcher viewing search results, I want to see all dates and date
  ranges associated with a dataset so that I can quickly see what time periods are covered.
* As a researcher viewing search results, I want to be able to re-sort by title
  or date range to find a specific resource or recent studies
* As a researcher viewing search results, I see 10 results per page by default,
  but can choose to view 25, 50, or 100 results so that it doesn't take as long
  to page through a large result set.
* As a researcher viewing search results, I want to see results sorted by relevance
  by default.
* As a researcher I want to view a list of results that includes title, summary,
  keywords, and relevance score so I can better determine which results I want
  to look at in detail.
* As a researcher viewing search results, I want an easy, visual indicator of
  the relevancy of results so that I can make a better judgement about which results to view.
* As a researcher, I want to view partial records from the search results page
  so I can determine records I want to use without leaving the search results page.
* As a researcher searching on multiple criteria, I want to specify terms
  in addition to keywords, for any of the following: title, summary, and/or source,
  so that I can find specific resources.
* As a researcher performing a search, I want to search by geographic location,
  including nation, country, and/or economic entity to tailor results to a specific
  geographic location or groups of locations.
* As a researcher, I want to use use grouping and boolean operators in the keyword
  search input to look for combinations of terms or search for different terms
  for the same concept, in order to find more precise results.
* As a researcher, I want to be able to jump anywhere within the full set of
  search results, so that I can easily get to any of the content.
* As a researcher entering search criteria, I want to be able to reset criteria
  to revert to the default before searching.
* As a researcher performing a search, I want to search by a single date where the
  date coverage includes that date, before or after a given date, or using a date
  range so that I can locate results by a specific time period.
* As a researcher, I want to be able to sort search results by either the beginning
  of the date coverage or the end, so that I can quickly determine which datasets
  have either the earliest coverage or the most recent coverage.
* As a researcher, I want to browse lists of newly purchased data collections so
  I can be aware of them.
* As a researcher, I want search results to be displayed in a way that makes the
  results easy to scan in order to find pertinent information more quickly.
* As a researcher viewing a single record, I want to be able to click on any of
  the associated keywords or topics to find other resources with those terms,
  in order to find related materials.
* As a researcher, I want to browse by topic so I can ensure I search broadly
  enough to find all relevant data.
* As a researcher, I want to be able to search by just date ranges alone, so that
  I can quickly determine what data are available for a particular time period.
* As a researcher, I want to browse by geographic region so I can see all data
  available for groups of countries, in addition to those available for individual
  countries.

Record Display
^^^^^^^^^^^^^^

* As a researcher, I want to display the full record associated with a search
  result, inc. summary, title, topic, date range, and geography so I can gain a
  comprehensive sense of the data.
* As a researcher viewing a single record, I want to be able to click on any of
  geographical coverage terms to find other resources which also cover those
  locations, in order to find related materials.
* As a researcher viewing a single record, I want to see a map of the locations
  covered by the data, so I can get a visual sense of the geographical coverage.
* As a researcher, when viewing the full record, I will know where to locate the
  data (online or in a physical location) so I can make use of the data.

Data Management
^^^^^^^^^^^^^^^

* As a data services staff member, I need to update and add newly purchased data
  collections to keep things up to date and better advertise the resources available.
* As a data services staff member, when I add or update data collection records,
  I want ICPSR topic headings mapped to a broader set of predefined topics so
  that users can browse by a curated list of topics.
* As a data services staff member, when I add or update data collection records,
  I want time period dates cleaned up so that users can consistently search content by dates.
* As a data services staff member, when I add or update data collection records,
  I want geographical terms categorized or coded, in order to provide users with
  better ways to browse and view datasets geographically.

Theme, branding, etc.
^^^^^^^^^^^^^^^^^^^^^

* Automated tools, such as search engine robots or site crawlers, can find

ddisearch/ddi/forms.py
# file ddisearch/ddi/forms.py
#
# Copyright 2014 Emory University
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#    http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from django import forms


class TelephoneInput(forms.TextInput):
    'HTML5 telephone input (prompt for numeric entry)'
    input_type = 'tel'

class SearchOptions(forms.Form):
    '''Base form with common options for search and browse result pages
    (items per page and sort)'''
    #: options for number of items per page
    PER_PAGE_OPTIONS = (10, 25, 50, 100)
    page_choices = [(d, d) for d in PER_PAGE_OPTIONS]
    #: per page input (:class:`~django.forms.ChoiceField`, defaults to 10)
    per_page = forms.ChoiceField(label='Results per page',
        choices=page_choices, initial=10)
    #: base options for sorting items
    SORT_OPTIONS = ['title', 'date (recent)', 'date (oldest)']
    sort_choices = [(d, d) for d in SORT_OPTIONS]
    #: sort input (:class:`~django.forms.ChoiceField`, defaults to title)
    sort = forms.ChoiceField(label='Sort by',
        choices=sort_choices, initial='title')

class KeywordSearch(SearchOptions):
    'Main DDI search form, with keyword and other field searches'
    #: keyword or phrase input (:class:`~django.forms.CharField`)
    keyword = forms.CharField(required=False, widget=forms.TextInput(attrs={'class':'form-control'}),
        help_text='One or more keywords; can include wildcards * and ?, and exact phrases in quotes.')

    # inherit per-page options and input
    # extend sort options to include relevance
    #: keyword sort options - includes relevance
    SORT_OPTIONS = ['relevance'] + SearchOptions.SORT_OPTIONS
    sort_choices = [(d, d) for d in SORT_OPTIONS]
    #: sort input (:class:`~django.forms.ChoiceField`, defaults to relevance)
    sort = forms.ChoiceField(label='Sort by',
        choices=sort_choices, initial='relevance')
    #: title search input (:class:`~django.forms.CharField`)
    title = forms.CharField(required=False,
        help_text='Search by title (keywords or phrase)',
        widget=forms.TextInput(attrs={'placeholder':'Search by title', 'class': 'form-control'}))
    #: record summary search input (:class:`~django.forms.CharField`)
    summary = forms.CharField(required=False,
        help_text='Search by summary (keywords or phrase)',
        widget=forms.TextInput(attrs={'placeholder':'Search by summary', 'class': 'form-control'}))
    #: source search input (:class:`~django.forms.CharField`)
    source = forms.CharField(required=False,
        help_text='Search by source organization or principial investigator (keywords or phrase)',
        widget=forms.TextInput(attrs={'placeholder':'Search by source', 'class': 'form-control'}))
    #: geographical location search input (:class:`~django.forms.CharField`)
    location = forms.CharField(required=False,
        help_text='Search by geographic location, including nation, country, or economic entity',
        widget=forms.TextInput(attrs={'placeholder':'Search by location', 'class': 'form-control'}))

    # NOTE: could use min_value and max_value here, if we have reasonable values
    # default widget is number input; we can switch to text input if helpful
    # this will nned to be a tel input to allow for devices to display the number pad automatically
    # also would be good to add the class form-control to the <input> element
    #: start date search input (:class:`~django.forms.IntegerField`)
    start_date = forms.IntegerField(required=False,
        help_text=''''Search by start year of time period covered by data;
        use with end date to specify a range or single year''',
        widget=TelephoneInput(attrs={'class': 'form-control', 'placeholder': 'Start year'}))
    #: end date search input (:class:`~django.forms.IntegerField`)
    end_date = forms.IntegerField(required=False,
        help_text='Search by end date; use with start date to specify a range or single year',
        widget=TelephoneInput(attrs={'class': 'form-control', 'placeholder': 'End year'}))

    _adv_fields = ['title', 'summary', 'source', 'location']

    search_fields = ['keyword', 'start_date', 'end_date'] + _adv_fields
    # fields that can contain search term; at least one of these is required
    # for validation; all combined generate a list of all search terms

    @property
    def advanced_fields(self):
        'list fields that are considered part of the "advanced" search'
        return [self[f] for f in self._adv_fields]

    def clean(self):
        'custom validation to ensure at least one search field is populated'
        cleaned_data = super(KeywordSearch, self).clean()
        # any one search term is sufficient, but at least one is required
        if not any([cleaned_data.get(f, None) for f in self.search_fields]):
            raise forms.ValidationError('You must enter search terms for at least one of %s or %s' % \
                                        (', '.join(self.search_fields[:-1]),
                                         self.search_fields[-1]))
        return cleaned_data

    @property
    def all_search_terms(self):
        'string of all search terms present in any keyword/phrase search fields'
        try:
            cleaned_data = self.clean()
            return ' '.join([cleaned_data.get(f) for f in self.search_fields
                             if f in cleaned_data and cleaned_data[f]
                                                  and not '_date' in f])
            # skip date fields when generating search terms
        except Exception as e:
            print 'exception: %s' % e
            # if form is not valid, then there are no search terms
            return ''


class BrowseTerms(SearchOptions):
    '''Form for browse by topic or keyword; uses base search options to
    include per_page and sort options, and has optional, hidden fields
    to pass browse value for current topic or keyword.'''
    #: topic search input (:class:`~django.forms.HiddenField`)
    topic = forms.CharField(required=False, widget=forms.HiddenInput)
    #: keyword search input (:class:`~django.forms.HiddenField`)
    keyword = forms.CharField(required=False, widget=forms.HiddenInput)
